vpc/service/internal_apis.go

```diff
@@ -407,3 +410,234 @@ func (vpcService *vpcService) doDetachBranchNetworkInterface(ctx context.Context
+	}
+
+	// 2. Get Default security group ID of the VPC to use in place of the reset SG ID
+	logger.G(ctx).Debug("2. Get Default security group ID of the VPC to use in place of the reset SG ID ", sgToDelete)
+	var defaultSecurityGroupID, region, account string
+	defaultSgIDS := make([]string, 0)
+	var resetSgSession *ec2wrapper.EC2Session
+	rows, err = resetSgTx.QueryContext(ctx, `
+SELECT region, account, group_id FROM security_groups WHERE  group_name='default' and account IN (SELECT account FROM security_groups WHERE group_id = $1) limit 1
+		`, sgToDelete)
+	if err == sql.ErrNoRows {
+		err = fmt.Errorf("Security group %s does not exist: %w ", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &empty.Empty{}, err
+	}
+	if err != nil {
+		err = fmt.Errorf("Could not query database for VPC ID of %s: %w ", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &empty.Empty{}, err
+	}
+	for rows.Next() {
+		err := rows.Scan(&region, &account, &defaultSecurityGroupID)
+		if err != nil {
+			err = fmt.Errorf("Could not get VPC ID of %s: %w ", sgToDelete, err)
+			tracehelpers.SetStatus(err, span)
+			return &empty.Empty{}, err
+		}
+
+		if err != nil {
+			err = fmt.Errorf("Could not get EC2 session for branch ENI: %w", err)
+			tracehelpers.SetStatus(err, span)
+			return &empty.Empty{}, err
+		}
+	}
+
+	if len(defaultSecurityGroupID) == 0 {
+		err = fmt.Errorf("Security group %s does not exist: %w ", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &empty.Empty{}, err
+	}
+
+	//Terminate transaction before calling the ec2 API and doing an UPDATE
+	err = resetSgTx.Commit()
+	if err != nil {
+		err = errors.Wrap(err, "Could not commit transaction")
+		tracehelpers.SetStatus(err, span)
+		return nil, err
+	}
+
+	resetSgSession, err = vpcService.ec2.GetSessionFromAccountAndRegion(ctx, ec2wrapper.Key{Region: region, AccountID: account})
+	if err != nil {
+		err = fmt.Errorf("Unable to create an ec2 session to reset SG %s : %w ", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &empty.Empty{}, err
+	}
+
+	defaultSgIDS = append(defaultSgIDS, defaultSecurityGroupID)
+
+	//Start new transaction
+	resetSgTx, err = beginSerializableTx(ctx, vpcService.db)
+	if err != nil {
+		err = errors.Wrap(err, "Unable to begin serializable transaction")
+		tracehelpers.SetStatus(err, span)
+		return &empty.Empty{}, err
+	}
+	defer func(tx *sql.Tx) {
+		_ = tx.Rollback()
+	}(resetSgTx)
+
+	//4. Reset the SG Id to default
+	logger.G(ctx).Debug("4. Reset the SG Id to default %s", sgToDelete)
+	rows, err = resetSgTx.QueryContext(ctx,
+		"UPDATE branch_enis SET security_groups = ARRAY[$1],dirty_security_groups=true WHERE ARRAY[$2] <@ security_groups returning branch_eni",
+		defaultSecurityGroupID, sgToDelete)
+	if err != nil {
+		err = fmt.Errorf("Cannot mark security groups as dirty for %s : %w", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &empty.Empty{}, err
+	}
+
+	var enisWithSgToUpdate []string
```

Comment:
On a per ENI basis you should:

1. Open a traditional TXN
2. SELECT FOR NO KEY UPDATE on the branch eni
3. Alter it in AWS
4. Update the dirty flag to false.

Reply:
On step 2, make sure the dirty flag is still true, and only update it to the values in the DB.